### ISSUE TITLE
AbstractFunctionValues

### ISSUE THREAD

**Issue by KnutAM** [Ferrite-FEM/Ferrite.jl#837]
Adding this here as a reminder and in case the discussion [here](https://julialang.slack.com/archives/C01R8JU271D/p1699179515802179) enters the Slack hole.

To be considered after #764 and #798 

Currently, there are `PointValues` and `PointValuesInternal` which allow updating the quadrature point location. `PointValues` just contains a `CellValues` whose quadrature points are updated during `reinit!`, and `PointValuesInternal` has only the shape values in the reference domain (which for identity mapping are the same as in the spatial domain). 

If we introduce `AbstractFunctionValues` or another way to have multiple variants of `FunctionValues` (see #764), which each have a different number of derivatives, we can use these to construct `AbstractValues` which are able to give only the value (current `PointValuesInternal` only), value and 1st derivative (current CellValues, FaceValues, PointValues), or value + 1st and 2nd order derivatives (currently not implemented, but [requested](https://julialang.slack.com/archives/C01R8JU271D/p1697461625968219)). 
(This would also merge `PointValuesInternal` into `PointValues` and make it possible to expose this API for use in e.g. `FerriteViz.jl`)

**Comment by KnutAM** [comment]
The merged `FunctionValues` has the `DiffOrder` type parameter and `PointValues` and `PointValuesInternal` are merged. The API for requesting a different `DiffOrder` is not public/documented yet, since we might want to change the API after introducing the option of adding `DiffOrder` higher than 1. 

With that generalization, I think this issue can be closed. If `AbstractFunctionValues` comes up again but for a different reason, that would be a separate issue.
